### PR TITLE
Psi compacting

### LINKED ISSUE #949: Multiple cases together stop parsing
**Desktop (please provide environment info):**
 - OS: Windows 10
 - IDE: intellij 2024.1.2
 - Plugin Version 1.34

**Describe the problem**

When two cases are next to each other parsing stops so the code folding no longer occurs.

**Please provide full details to reproduce the problem**

```d
switch(expr) {
    case a:
    case b:
       break;
    default:
        break;
}
```

### PR DESCRIPTION
**Pull request by etienne02**
Sorry for the size of the PR but it’s difficult (or impossible) to split it into multiple. To review it, you can review commit by commit but it still hard to follow the changes (the firsts commits have a structure, but at the end I mixed fixes and new changes so the global view may be better).
Note that I did not changed the test cases (except some parsing but the structure was invalid, example: it’s not allowed to put version specification in statement, so in unittest)

**This PR aims to:**
1. Clean the parsing (make his structure more clear and more robust, no more "mixin" like method)
2. Fix some parsing errors, the previous parsing wrongly parsed some stuff, now it’s more compliant to the official specs
3. Reduce the psi depth, collapsing some nodes and introducing some interfaces (declaration, statement, expression). Especially no node is created for expression if no expression of this type is here (no AddExpression node if no `+` in the expression)

To do this some Psi node were removed and other were added (also some were renamed to follow more the official specs naming). I also removed the PsiLightParsing stuff as it was a lot of lines of code for nothing (it’s always parsed al Module level only).
The effect of this is that the Psi is much more clean and clear to traverse and understand. This will help a lot for features like semantic analysis and resolve.

As a side effect, some bug/regressions are fixed (it was not intended by this PR but it’s a positive effect)
Fixes #949 #948 and #811 (do not close them, tests need to be added and will be made in another PR)

Some regressions are expected (especially in resolve, I know some cases that does not resolve anymore but they are not covered by test)

More improvements can still be made and will probably be made (but in other PRs)
Possible improvements points:
1. Error recovering during parsing:
   currently the parser always put itself in situation where it’s able to continue and so it never end with infinite loop. But it’s really generic and some better error messages and recovery code can be made to help the user to understand the errors.
2. Make StatementBlock lazy (see Java for this point), this could improve the indexing speed (as statements won’t be parsed in detailed during indexing.

My continuation plan to improve the state of the plugin
1. Remove Identifier Node (and especially fact that it’s NamedNode) no other plugin is doing that and we also should not do that
2. Refactor/Reimplement the resolve algorithm:
  A lot of improvements can be made for resolve, but the current algorithm is not clear enough. A better structure and algorithm should be implemented to make it really powerful. Also add support to detect fields/functions of an element (a.b when b is the field of struct A).
3. Better and more useful indexes

Note that the PSI structure is still not perfect and may change a bit in the future but not that much as with this PR.

### CODE REVIEW

**Review by SingingBush — APPROVED** [review]
thanks, this looks good but I've not checked the branch out for any local testing. Do you want to continue on this branch or get this squashed & merged into develop as it is before making further changes? I will spend some time trying out these changes but if you have more planned then perhaps I should wait

**Comment by etienne02** [comment]
I rather having it merged and squashed. This PR is big enough. This does not prevent me for doing further changes (so you can take time to check and test) but I’ll create other PRs for the other work I’ll do
